aiida/tools/importexport/migration/v06_to_v07.py

```diff
@@ -0,0 +1,99 @@
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Migration from v0.6 to v0.7, used by `verdi export migrate` command.
+
+The migration steps are named similarly to the database migrations for Django and SQLAlchemy.
+In the description of each migration, a revision number is given, which refers to the Django migrations.
+The individual Django database migrations may be found at:
+
+    `aiida.backends.djsite.db.migrations.00XX_<migration-name>.py`
+
+Where XX are the numbers in the migrations' documentation: REV. 1.0.XX
+And migration-name is the name of the particular migration.
+The individual SQLAlchemy database migrations may be found at:
+
+    `aiida.backends.sqlalchemy.migrations.versions.<id>_<migration-name>.py`
+
+Where id is a SQLA id and migration-name is the name of the particular migration.
+"""
+# pylint: disable=invalid-name
+from __future__ import division
+from __future__ import print_function
+from __future__ import absolute_import
+
+from aiida.tools.importexport.migration.utils import verify_metadata_version, update_metadata
+
+
+def migration_data_migration_legacy_process_attributes(data):
+    """Apply migration 0040 - REV. 1.0.40
+    Data migration for some legacy process attributes.
+
+    Attribute keys that are renamed:
+
+    * `_sealed` -> `sealed`
+
+    Attribute keys that are removed entirely:
+
+    * `_finished`
+    * `_failed`
+    * `_aborted`
+    * `_do_abort`
+
+    Finally, after these first migrations, any remaining process nodes that still do not have a sealed attribute, have
+    it set to `True`. Since there can be no active processes in an export file, this is not a problem.
+
+    :raises `~aiida.common.exceptions.IntegrityError`: if a Node, found to have attributes, cannot be found in the list
+        of exported entities.
+    """
+    from aiida.common.exceptions import IntegrityError
+
+    attrs_to_remove = ['_sealed', '_finished', '_failed', '_aborted', '_do_abort']
+
+    for node_pk, content in data['node_attributes'].items():
+        try:
+            if data['export_data']['Node'][node_pk]['node_type'].startswith('process.'):
+                # Remove attributes
+                for attr in attrs_to_remove:
+                    content.pop(attr, None)
+
+                # Set `'sealed' = True`, if 'sealed' does not already exist
+                if 'sealed' not in content:
+                    content['sealed'] = True
```

Comment:
I know this has been already discussed, but is there a real reason not to put the *same* logic checking the state (as in the migrations) also here?

Reply:
One would never have been able to export a running Node, I presume? So no reason to check state.

Reply:
Well, I'm pretty sure that even in the current develop version it *is* possible, that's why I'd prefer to do the same check. The goal of the data migration is to mirror the SQL migrations, so I'd rather go for explicitly implementing the same logic (that, BTW, isn't so complex in this case) rather than using assumptions that might be not true in some cases

Reply:
This is fine, I will implement this.
My only question is now - what should the migration do for the case where a Node is in an active state? Since we should not allow these ProcessNodes to be imported, I don't want the migration to create a valid, importable archive.
For this case, shouldn't we acknowledge the difference between the archive and the database and simply raise, stating that the archive is corrupt - print out the affected ProcessNodes, and have them contact us? Or perhaps even (dangerous) give them the option to override the process_state to finished and "seal" the Node...?

Reply:
> The goal of the data migration is to mirror the SQL migrations, (...)

Yes, I agree, but only to the extent of what is "allowed" to export/import.
If we agree _only_ sealed ProcessNodes can be exported from here on out (something that this PR is implementing), then we should consider what to do when migrating exported ProcessNodes that could clearly never have been sealed when originally exported (i.e., they have an active `process_state`).
Personally, I think the archive should be marked as corrupted, and they should contact us to sort it out (as I wrote above). But perhaps we cannot come up with any case, where it is _not_ okay to just change the `process_state` to `"finished"`, set the `sealed` attribute and be done with it?

In any case - I am sorry for writing two almost similar comments here, I just felt like the first one didn't properly explain my position, but perhaps I was wrong :)